apps/io-wallet-user-func/src/android-key-attestation.ts

```diff
@@ -0,0 +1,127 @@
+} & WalletInstanceValid;
+
+const hasAndroidCertificatesChain = (
+  walletInstance: WalletInstanceValid,
+): walletInstance is WalletInstanceValidWithAndroidCertificatesChain =>
+  !!walletInstance.deviceDetails &&
+  walletInstance.deviceDetails.platform === "android" &&
+  !!walletInstance.deviceDetails.x509Chain;
+
+const validateKeyAttestationChain = async (
+  walletInstance: WalletInstanceValid,
+  attestationServiceConfiguration: AttestationServiceConfiguration,
+) => {
+  // Check if device details exist and the platform is Android with a valid x509 chain. For iOS is not necessary.
+  if (hasAndroidCertificatesChain(walletInstance)) {
+    const x509Chain = walletInstance.deviceDetails.x509Chain.map(
+      (cert) => new X509Certificate(cert),
+    );
+
+    // Validate issuance and expiration of the certificates
+    validateIssuance(
+      x509Chain,
+      attestationServiceConfiguration.googlePublicKey,
+    );
+    // Validate revocation of the certificates using the provided CRL URL
+    await validateRevocation(
+      x509Chain,
+      attestationServiceConfiguration.androidCrlUrl,
+      attestationServiceConfiguration.httpRequestTimeout,
+    );
+  }
+};
+
+const validateAttestedKeyAndRevokeIfNecessary: (
+  walletInstance: WalletInstanceValid,
+) => RTE.ReaderTaskEither<
+  {
+    attestationServiceConfiguration: AttestationServiceConfiguration;
+    telemetryClient: appInsights.TelemetryClient;
+    walletInstanceRepository: WalletInstanceRepository;
+  },
+  Error,
+  void
+> =
+  (walletInstance: WalletInstanceValid) =>
+  ({
+    attestationServiceConfiguration,
+    telemetryClient,
+    walletInstanceRepository,
+  }) =>
+    pipe(
+      TE.tryCatch(
+        () =>
+          validateKeyAttestationChain(
+            walletInstance,
+            attestationServiceConfiguration,
+          ),
+        E.toError,
+      ),
+      // If the validation fails, revoke the user's wallet instance
+      TE.alt(() =>
+        pipe(
+          revokeUserWalletInstances(walletInstance.userId, [walletInstance.id])(
+            {
+              walletInstanceRepository,
+            },
+          ),
+          TE.map(() =>
+            telemetryClient.trackEvent({
+              name: "REVOKED_WALLET_INSTANCE_FOR_INVALID_CERTIFICATE",
+              properties: {
+                fiscalCode: walletInstance.userId,
+                walletInstanceId: walletInstance.id,
+              },
+            }),
+          ),
+        ),
+      ),
+    );
+
+export const checkWalletInstancesAttestedKeyRevocation: RTE.ReaderTaskEither<
+  {
+    attestationServiceConfiguration: AttestationServiceConfiguration;
+    telemetryClient: appInsights.TelemetryClient;
+    walletInstanceRepository: WalletInstanceRepository;
+  },
+  Error,
+  void
+> = (params) =>
+  pipe(params, getAllValidWalletInstances, (generator) =>
+    TE.tryCatch(async () => {
+      for await (const result of generator) {
+        pipe(
```

Comment:
if you don't await here, in order to process tasks in parallel, you have to await the generated promises somewhere else using promise.all. Another option is to collect Tasks using Array.from(generator).map(...) and then call TE.sequenceArray(tasks)

Something like

```ts
export const checkWalletInstancesAttestedKeyRevocation: RTE.ReaderTaskEither<
  {
    attestationServiceConfiguration: AttestationServiceConfiguration;
    telemetryClient: appInsights.TelemetryClient;
    walletInstanceRepository: WalletInstanceRepository;
  },
  Error,
  void
> = (params) =>
  pipe(
    params,
    getAllValidWalletInstances,
    TE.map(generator =>
      // Collect all TaskEithers into an array
      Array.from(generator).map(result =>
        pipe(
          result,
          E.fold(
            () => TE.right(undefined),
            (walletInstance) =>
              pipe(
                params,
                validateAttestedKeyAndRevokeIfNecessary(walletInstance)
              )
          )
        )
      )
    ),
    TE.chain(tasks =>
      // Execute all TaskEithers in parallel
      TE.sequenceArray(tasks).map(() => undefined)
    )
  );
```

Reply:
I thought about it but I don't like this approach because of the memory usage. Consider that there could be millions of active Wallet Instances and having such a large array in memory worries me. The approach used instead of fire and forget should solve this problem.

Reply:
As tasks run in parallel, you either have some way to limit concurrency otherwise you get the same issues when using array.from, since every tasks starts just after the previous one without wait (memory usage will be the same if not more)